### ISSUE TITLE
Prevent focus loss when deleting all blocks

### ISSUE THREAD

**Issue by jasmussen** [WordPress/gutenberg#9626]
Extracted from https://github.com/WordPress/gutenberg/pull/8961#issuecomment-418682481.

**Problem:** If you delete all blocks on a post or page, focus is lost. 

Steps to reproduce:

1. Insert a bunch of blocks. Try the demo content. 
2. Press ⌘A to select all blocks (⌘A twice if your caret is in text first)
3. Press del to delete those blocks. 

Notice how focus is lost (it's actually on the `body` element).

**Propost solution:**

When all blocks are deleted, simply set focus on the first appender:

<img width="731" alt="screen shot 2018-08-31 at 11 34 29" src="https://user-images.githubusercontent.com/1204802/44905274-dbe3f880-ad11-11e8-9f49-86e71dca2f38.png">

CC: @talldan

**Comment by aduth** [comment]
As far as implementation, the obvious path seems to be that when the last block is removed, we insert and focus the default block. However, this raises a few questions:

- What happens when the only thing left is the empty paragraph and the user wants to remove it?
   - [Related Slack discussion](https://wordpress.slack.com/archives/C02QB2JS7/p1533663284000301) ([link requires registration](https://make.wordpress.org/chat/))
   - If we say that the default block is inserted when the last of all blocks are removed, that means that if I'm already in the default empty block and remove it, I'll be forever trapped in this repeated cycle of default block insertion.
   - Recommendation: Maybe that's okay. And maybe we should change the serialization step to account for the fact that a block which only includes the "empty default block" should not include any content (rather than the empty paragraph tag with delimiters).
- Is a similar behavior needed for blocks within a nested context?
   - Recommendation: It's not as easy to account for the issue described in the first point, so while technically inconsistent it might be better to reserve this behavior for top-level blocks removal only.
- What if the default block is not the paragraph block? A custom post type could, for example, assign `core/image` as the "default" block type. This is technically supported.
   - Recommendation: There may be a bigger question on the interactions of non-paragraph default blocks, but in other cases we apply the behavior consistently as well (e.g. "default block appender") so it should be a non-issue.

**Comment by jasmussen** [comment]
> If we say that the default block is inserted when the last of all blocks are removed, that means that if I'm already in the default empty block and remove it, I'll be forever trapped in this repeated cycle of default block insertion.

That is an interesting point. 

> Recommendation: Maybe that's okay. And maybe we should change the serialization step to account for the fact that a block which only includes the "empty default block" should not include any content (rather than the empty paragraph tag with delimiters).

I like this.

A Plan B could be to set focus on the title block instead. A plan C could be to set focus on the editing canvas, so a single press of Tab sets focus on the title. Right now focus is on `body`, so you have a great deal to tab through to get back to where you were.

> Is a similar behavior needed for blocks within a nested context?

Depends? I think the key is to ensure focus is set on a sensible place when a block is deleted. Could be _the preceeding block_, which suggests focus could be set on the title if all blocks are deleted.

**Comment by aduth** [comment]
As of #9808, a post containing only an unmodified default block (paragraph) is now saved as empty.

This enables #9977, which implements the behavior proposed here in ensuring always a default empty block is inserted into the root of the editor upon removal / replacement.
